src/VecSim/algorithms/hnsw/hnsw_tiered.h

```diff
@@ -73,3 +108,90 @@ class TieredHNSWIndex : public VecSimTieredIndex<DataType, DistType> {
+    auto **jobs = array_new<AsyncJob *>(1);
+    jobs = array_append(jobs, job);
+    this->SubmitJobsToQueue(this->jobQueue, (void **)jobs, 1);
+    array_free(jobs);
+}
+
+/******************** Index API ****************************************/
+
+template <typename DataType, typename DistType>
+TieredHNSWIndex<DataType, DistType>::TieredHNSWIndex(HNSWIndex<DataType, DistType> *hnsw_index,
+                                                     TieredIndexParams tieredParams)
+    : VecSimTieredIndex<DataType, DistType>(hnsw_index, tieredParams),
+      labelToInsertJobs(this->allocator), idToRepairJobs(this->allocator),
+      idToSwapJob(this->allocator) {
+    this->UpdateIndexMemory(this->memoryCtx, this->getAllocationSize());
+}
+
+template <typename DataType, typename DistType>
+TieredHNSWIndex<DataType, DistType>::~TieredHNSWIndex() {
+    // Delete all the pending insert jobs.
+    for (auto jobs : this->labelToInsertJobs) {
+        for (auto *job : jobs.second) {
+            delete job;
+        }
+    }
+}
+
+template <typename DataType, typename DistType>
+size_t TieredHNSWIndex<DataType, DistType>::indexSize() const {
+    return this->index->indexSize() + this->flatBuffer->indexSize();
+}
+
+template <typename DataType, typename DistType>
+size_t TieredHNSWIndex<DataType, DistType>::indexCapacity() const {
+    return this->index->indexCapacity() + this->flatBuffer->indexCapacity();
+}
+
+template <typename DataType, typename DistType>
+size_t TieredHNSWIndex<DataType, DistType>::indexLabelCount() const {
+    return this->flatBuffer->indexLabelCount() + this->index->indexLabelCount();
+}
+
+template <typename DataType, typename DistType>
+int TieredHNSWIndex<DataType, DistType>::addVector(const void *blob, labelType label,
+                                                   bool overwrite_allowed) {
+    /* Note: this currently doesn't support overriding (assuming that the label doesn't exist)! */
+    this->flatIndexGuard.lock();
+    if (this->flatBuffer->indexCapacity() == this->flatBuffer->indexSize()) {
+        this->flatBuffer->increaseCapacity();
+    }
+    idType new_id = this->flatBuffer->indexSize();
+    this->flatBuffer->addVector(blob, label, false);
+    AsyncJob *new_insert_job = new (this->allocator)
+        HNSWInsertJob(this->allocator, label, new_id, executeInsertJobWrapper, this);
+    // Save a pointer to the job, so that if the vector is overwritten, we'll have an indication.
+    if (this->labelToInsertJobs.find(label) != this->labelToInsertJobs.end()) {
+        // There's already a pending insert job for this label, add another one (without overwrite,
+        // only possible in multi index)
+        assert(this->index->isMultiValue());
+        this->labelToInsertJobs.at(label).push_back((HNSWInsertJob *)new_insert_job);
+    } else {
+        auto new_jobs_vec = vecsim_stl::vector<HNSWInsertJob *>(1, (HNSWInsertJob *)new_insert_job,
```

Comment:
this is using a copy constructor, right?

Reply:
Yep. I guess I can avoid the copy and just create the vector instead.